### ISSUE TITLE
Support for Aventra MyEID PUT DATA: INITIALISE PIV EMULATION command

### ISSUE THREAD

**Issue by minfrin** [OpenSC/OpenSC#1754]
### Problem Description

According to https://webservices.aventra.fi/wordpress/wp-content/downloads/MyEID%20PKI%20JavaCard%20Applet%20Reference%20Manual%202-3-0.pdf

> The MyEID4 card can emulate a PIV/CIV card by mapping the ISO 7816-15 (PKCS#15) structure to the PIV/CIV command interface. The PIV/CIV emulation can be switched on and off with a single command.

Does opensc support the issuing of this single command?

If not, where in opensc would be the best place for such a command? piv-tool?

### Proposed Resolution

Add support for PUT DATA: INITIALISE PIV EMULATION to piv-tool.

**Comment by dengert** [comment]
I would suggest that the `piv-tool.c` is not the place for this command. The `piv-tool.c` uses the PIV driver `card-piv.c` to issue PIV commands defined in the PIV standards "NIST sp800-73-*" to write data objects and generate keys on the card. Just enough to personalize test cards. It only works if the PIV AID can be selected. 

I would suggest the the command be issued by some MyEID specific program that has access to 
the PKCS\#15 structures.  This is common practice for OpenSC:  cardos-tool.c. egk-tool.c. iasecc-tool.c,  openpgp-tool.c,   westcos-tool.c, cryptoflex-tool.c,  gids-tool.c,  netkey-tool.c, dnie-tool.c,        goid-tool.c,  npa-tool.c,  piv-tool.c , has  sc-hsm-tool.c.


The PIV NIST sp800-87-3 standards leave it up to PIV card vendors to supply additional commands not covered in the standards for card initialization and personalization. Card vendors would then provide the commands to card management systems vendors, giving card vendors some competitive edge in selling their cards.   

I believe the command your are referring to is:

"5.3 PUT DATA: INITIALISE PIV EMULATION
The PUT DATA: INITIALISE PIV EMULATION command activates the PIV/CIV interface and
maps selected keys and certificates to corresponding PIV objects.
Before issuing this command, the keys and certificates to be used must be loaded on the
card using MyEID command interface. Mapping objects is optional. For example, if only
authentication key and certificate are needed, other FIDs can be set to 0000 in the PIV
Initialisation Parameters structure."

This has to be done before the SELECT AID for PIV can be selected. `piv-tool.c` and `card-piv.c` depended on the AID being selected. 

I am concerned with the other half of the paragraph you quoted above"

"The MyEID4 is not fully PIV/CIV compliant; the aim is to
emulate the functionality to the extent of interoperability on command interface level,
and to allow using MyEID cards in PIV compatible devices, in situations where native
MyEID middleware cannot be installed."
 
There are other cards that are 'not  PIV/CIV compliant" i.e. they have not gone thru NIST certification.  This has caused problems especially when the card/token can support multiple applets on the card. These include Yubikey with PIV and OpenPGP, PIV/CAC cards.  Yubico provides the Yubikey tools. Governments issue PIV/CAC and PIV cards using card management systems. OpenSC does not support personalanization.

**Comment by frankmorgner** [comment]
In OpenSC, you can send any apdu you want via `opensc-tool --apdu`; feel free to add this to the [MyEID documentation](https://github.com/OpenSC/OpenSC/wiki/Aventra-MyEID-PKI-card).

But, AFAIK, OpenSC has native support for myeid, so there's no need to switch to PIV...

**Comment by dengert** [comment]
I agree with Frank. On what platform would you need to use the PIV applet? 

 If the goal is to use the card as PIV in systems with no modifications, it should be noted a PIV CHUID object may be needed. piv-tool can write a CHUID, but OpenSC does not provide a way to format a CHUID.  (Yubikey tools do.)  (OpenSC used the CHUID to gererate a "serial" number if present. Windows builtin PIV driver also use the CHUID.)

It is not clear once the PIV applet is activated, is it the default? OpenSC's matches on ATRs, and/or on AID be doing SELECT AID commands. So the order of the drivers is important. I would suspect the MyEID driver would always be called before the PIV driver. 

Once the MyEID card is personalized, to turn on or off the PIV applet should require some authorization to the card. It is not clear how the "INITIALISE PIV EMULATION command" is protected. It should be, because it requires the object to be written to the card and one of the fields is an ACL.
So using just opensc-tool --apdu  should not work.

**Comment by minfrin** [comment]
My understanding is the MyEID card offers a read only PIV applet, which exposes the PKCS15 data.

All write preparation is done with PKCS15, and is supported by opens now.

The last step is to turn on PIV.

Everything else can be done with opensc, except for the last step.

**Comment by dengert** [comment]
You could still  use `opensc-tool --apdu` if done before doing a `pkcs15-init --finalize`.  as it looks no additional authorization is needed  before finalized. You would need to make make up the APDU using Table 66 based on the FIDs to add. Then do a finalize. 

You might want to ask Aventra to support writing of the  `INITIALISE PIV EMULATION` after a finalize is done. `pkcs15init/pkcs15-myeid.c` has their copyright. 

@hhonkanen   Do you have any ideas on this? 

I they don't do it,  pkcs15init would be the place to start could be used as the operation is to write out information as defined in the document.  `INITIALISE PIV EMULATION` is not a PIV command, it is a MyEID PUT DATA command to make the card emulate a PIV card when the PIV AID is selected.  

`card-myeid.c` has a `myeid_putdata` that will  issue a APDU of "00 DA P1 P2". the 
"PUT DATA: INITIALISE PIV EMULATION" uses "00 DA 01 50".  It is used via `myeid_card_ctl` with
`SC_CARDCTL_MYEID_PUTDATA`. 

`pkcs15init/pkcs15-myeid.c` has 2 calls to `sc_card_ctl(p15card->card, SC_CARDCTL_MYEID_PUTDATA, &data_obj);`  from `myeid_erase_card` and `myeid_create_pin` which set different P1 and P2 parameters in `struct sc_cardctl_myeid_data_obj`  

You could start with `myeid_create_pin` as an example, in the `sc_cardctl_myeid_data_obj`. 
What you need to do set P1 and P2 and set Data and DataLen to match the Table 66 in the document based on the PKCS15  FID you have written.  

Then figure out how to add another opt to the `sc_pkcs15init_operations`  and where to call it from  pkcs15init.

**Comment by minfrin** [comment]
> But, AFAIK, OpenSC has native support for myeid, so there's no need to switch to PIV...

MacOS out the box supports PIV.

We want to provision the cards using opensc, but use the cards on MacOS using native functionality.

**Comment by dengert** [comment]
I will leave the comments about MacOS and PIV to others...

But for a test you might try the trick (if it works)  above of doing the MyEID initialization without finalize, then use `opensc-tool --apdu "00:DA:01:50:14   80:....."` Where the 80 is the state from Table 66, and you have to figure out the reset of Table 66.  Then do the finalize.

**Comment by dengert** [comment]
I have 2 cards labeled "MyEID 4 PIV mode initialized" and "MyEID 4 uninitalized" from @hhonkanen  at  Aventra data 16.1.2017. These were for development, and I never got very far with them. 

So I am looking at them again. 

To verify that a card is a PIV card, two methods are used by the PIV driver. 
A test is made for the optional "Discovery object" which can be read without causing any problems.
That fails which is OK as this is optional.

Then a SELECT AID for the PIV is done and the results are tested to make sure it returned the PIV AID. (This is done as some cards can return a false positive by returning just the status '9000' or some data that does not have the AID.) 

The PIV driver does not like the response from SELECT AID as it appears to be missing the `61 39` TLV as expected from NIST approved cards. (This may be fixed it current cards.)

MyEID test card with PIV initialized:
```
./opensc-tool -c default -s "00 A4 04 00 09 A0 00 00 03 08 00 00 10 00 00"
Using reader with a card: SCM Microsystems Inc. SCR 355 [CCID Interface] 00 00
Sending: 00 A4 04 00 09 A0 00 00 03 08 00 00 10 00 00 
Received (SW1=0x90, SW2=0x00):
4F 06 00 00 10 00 01 00 79 08 4F 06 00 00 10 00 O.......y.O.....
01 00 50 18 4D 79 45 49 44 20 50 49 56 20 63 61 ..P.MyEID PIV ca
72 64 20 65 6D 75 6C 61 74 69 6F 6E             rd emulation
```

NIST DEMO card2:
```
./opensc-tool -c default -s "00 A4 04 00 09 A0 00 00 03 08 00 00 10 00 00"
Using reader with a card: SCM Microsystems Inc. SCR 355 [CCID Interface] 00 00
Sending: 00 A4 04 00 09 A0 00 00 03 08 00 00 10 00 00 
Received (SW1=0x90, SW2=0x00):
61 39 4F 0B A0 00 00 03 08 00 00 10 00 01 00 79 a9O............y
07 4F 05 A0 00 00 03 08 50 0E 49 44 2D 4F 6E 65 .O......P.ID-One
20 50 49 56 20 42 49 4F 5F 50 10 77 77 77 2E 6F  PIV BIO_P.www.o
62 65 72 74 68 75 72 2E 63 6F 6D 7F 66 08 02 02 berthur.com.f...
80 00 02 02 80 00                               ......
```
As expected the initialized card fails:
```
./opensc-tool -c default -s "00 A4 04 00 09 A0 00 00 03 08 00 00 10 00 00"
Using reader with a card: SCM Microsystems Inc. SCR 355 [CCID Interface] 00 00
Sending: 00 A4 04 00 09 A0 00 00 03 08 00 00 10 00 00 
Received (SW1=0x6A, SW2=0x82)
```

I will see how far I can get.

**Comment by dengert** [comment]
I can not get very far. The MyEID driver says `sc_connect_card: card info name:'MyEID 4.0.0', type:20001, flags:0x0, max_send/recv_size:255/256`

I can get around the SELECT AID issue with a hack. I expect this has already been fixed. The returned template needs to be wrapped in a tag "61" and the full AID need to be returned. See NIST sp800-73-3 parts 1 and 2 for response from SELECT AID. 

The test card came with a certificate and key:
```
pkcs11-tool -O --login
Using slot 0 with a present token (0x0)
Logging in to "John Smith".
Please enter User PIN: 
Private Key Object; RSA 
  label:      PIV AUTH key
  ID:         01
  Usage:      decrypt, sign, unwrap
  Access:     sensitive, always sensitive, never extractable, local
Public Key Object; RSA 2048 bits
  label:      PIV AUTH pubkey
  ID:         01
  Usage:      encrypt, verify, wrap
  Access:     none
Certificate Object; type = X.509 cert
  label:      Certificate for PIV Authentication
  subject:    DN: C=FI, CN=John Smith, O=Test Ltd
  ID:         01
...
```
But sign fails because the test card does not support data_chaining which is need when send the 256 byte padded hash + command data header to the card in two chunks. According `card-myeid.c` data_chaining is support version 4.5. 

The  Reference manual Ver. 2.3.0 says:  "25.3.2019  2.3.0 Added PUT DATA: Initialise PIV emulation" So I doubt my card supports it. 
 
This is an interesting problem. If I had an up to date card(s) I could take a look at it.

**Comment by hhonkanen** [comment]
I recommend trying @dengerts suggestion of sending the command to a fresh card using opensc-tool –apdu before issuing a pkcs15-init -F command. 

At first, check file id’s (FIDs) of keys and certificates using commands pkcs15-tool -c and pkcs15-tool -k, from a card containing the structure you are going to use.

You’ll get output like this: 

X.509 Certificate [Authentication certificate]
	Object Flags   : [0x02], modifiable
	Authority      : no
	Path           : 3f0050154331
	ID             : 45
	Encoded serial : 02 13 520000018C0C7BCEEBD92A742D00000000018C

X.509 Certificate [Signature certificate]
	Object Flags   : [0x02], modifiable
	Authority      : no
	Path           : 3f0050154332
	ID             : 46
	Encoded serial : 02 13 520000018D34E3584867D6995B00000000018D


Private RSA Key [Authentication key]
	Object Flags   : [0x03], private, modifiable
	Usage          : [0x37], encrypt, decrypt, sign, wrap, unwrap
	Access Flags   : [0x1D], sensitive, alwaysSensitive, neverExtract, local
	ModLength      : 2048
	Key ref        : 1 (0x01)
	Native         : yes
	Path           : 3f0050154b01
	Auth ID        : 01
	ID             : 45
	MD:guid        : 8852b23e-898d-865b-ceb6-faa2d95f662d

Private RSA Key [Signature key]
	Object Flags   : [0x03], private, modifiable
	Usage          : [0x37], encrypt, decrypt, sign, wrap, unwrap
	Access Flags   : [0x1D], sensitive, alwaysSensitive, neverExtract, local
	ModLength      : 2048
	Key ref        : 2 (0x02)
	Native         : yes
	Path           : 3f0050154b02
	Auth ID        : 01
	ID             : 46
	MD:guid        : d27a49e0-81e8-25ce-f3b3-c23e877de516

Take the FIDs from the “Path” field, they are the last four hex digits.

With the FIDs from my example card, the command would look like this:
opensc-tool --apdu "00:DA:01:50:14:80:11:1F:FF:4B:01:43:31:00:00:00:00:4b:02:43:32:00:00:00:00

If you always create the cards in the same way, you could first init the cards as described here: 
https://github.com/OpenSC/OpenSC/wiki/Aventra-MyEID-PKI-card
but issue the PUT DATA: INITIALISE PIN command before pkcs15-init -F
You can then load/generate the keys and certificates to the card later. 

I think the issue in SELECT AID response mentioned by @dengert is still not fixed and the card is therefore not compatible with OpenSC’s PIV driver. The PIV emulation has not been developed lately. However the cards work with Windows built-in PIV driver regardless of this. I don’t have experience on using the PIV emulation on Mac OS X, but we are interested to know if it works, if you can try it out.  I will check if we could do something for the SELECT AID issue in the near future.

**Comment by frankmorgner** [comment]
Closing this issue due to inactivity. Please re-open the ticket if more input is available.

**Comment by minfrin** [comment]
Keen to pick this up again on MacOSX.

If I am understanding this correctly, to make a MyEID card work on MacOS you would need to do the following:

- Initialise the MyEID card using the OpenSC tools in the normal way.
- Switch on PIV emulation by sending an APDU as described by @hhonkanen above.
- Finalise the MyEID card using the OpenSC tools in the normal way.
- Continue using PIV drivers on MacOS.

If this is true, then the outstanding work is to create an aventra-piv-tool command that does the switch on PIV emulation as described above and add this to opensc. Is this accurate?

Happy to buy a new card and try it out, my existing card is MyEID 3.3.3.

**Comment by minfrin** [comment]
Blank card, went through normal init process which installs two certs and two keys onto the card. Tried to issue the APDU and got the following:

```
Little-Net:~ minfrin$ opensc-tool --send-apdu "00:DA:01:50:14:80:11:1F:FF:4B:01:43:01:00:00:00:00:4B:02:43:03:00:00:00:00"
Using reader with a card: Gemalto Ezio Shield
Sending: 00 DA 01 50 14 80 11 1F FF 4B 01 43 01 00 00 00 00 4B 02 43 03 00 00 00 00 
Received (SW1=0x6A, SW2=0x86)
```

Assuming MyEID 3.3.3 is too old.

**Comment by hhonkanen** [comment]
MyEID 3.3.3 doens't support PIV emulation at all. Unfortunately even in the newest versions in market - 4.5.x - this feature is not really usable. I think it has been a mistake to include it in the reference manual before developing it further. This has caused some confusion. The PIV feature came in 4.0 and the intention was to enable operations like smart card logon with devices that have built in PIV drivers. And it worked at least with Windows 7 back then. But because our implementation was incomplete, it doesn't work in current MacOS versions or even in Windows anymore, because they require some things that are not implemented.

After long time of inactivity, PIV support is now under work again and we have been testing also MacOS. Our target is to define and document clearly which parts of the PIV spec are supported and verify by testing that the documented subset works in MacOS, Windows and with OpenSC's PIV driver. Please wait for the next MyEID version before buying them for PIV usage.

**Comment by dengert** [comment]
Please also have a look at  card-piv.c and "card_issues" that other PIV like cards have failed to implement correctly. (see lines  Loss of the login state is very important in an environment where multiple applications may be accessing the device, and testing the card by sending SELECT AID commands for other drivers.

Please also make sure there is a way to identify this is a MyEID card, and its version, incase it has additional issues. The response to the SELECT AID has a
tag for the vendor, which can be used. The historical bytes in the ATR can also be used. 

Also note 800-73-4 (and other versions) say in Part 2 1.1 Purpose:
"The goals are addressed by specifying a PIV data model, card edge interface, and application programming interface. Moreover, SP 800-73-4 enumerates requirements where the international integrated circuit card (ICC) standards [ISO7816] include options and branches. **The specifications go further by constraining implementers’ interpretations of the normative standards.** Such restrictions are designed to ease implementation, facilitate interoperability, and ensure performance, in a manner tailored for PIV applications.

One such restriction, is it does not require extended APDU, which also means it
must support command chaining for at least RSA 2048 crypto operations when using short APDUs.  And in Secure Messaging, its examples are all short and use command chaining. If using RSA 1024, it is not needed and maybe that is why Window 7 used to work.
